cupy/core/_gufuncs.py

```diff
@@ -0,0 +1,500 @@
+            keepdims (bool, optional):
+                If this is set to True, axes which are reduced over will be
+                left in the result as a dimension with size one, so that the
+                result will broadcast correctly against the inputs. This
+                option can only be used for generalized ufuncs that operate
+                on inputs that all have the same number of core dimensions
+                and with outputs that have no core dimensions , i.e., with
+                signatures like ``'(i),(i)->()'`` or ``'(m,m)->()'``.
+                If used, the location of the dimensions in the output can
+                be controlled with axes and axis.
+            out (cupy.ndarray): Output array. It outputs to new arrays
+                default.
+
+        Returns:
+            Output array or a tuple of output arrays.
+        '''
+
+        #  This argument cannot be used for generalized ufuncs
+        #  as those take non-scalar input.
+        # where = kwargs.pop('where', None)
+
+        dtype = args[0].dtype
+        for arg in args:
+            ret_dtype = numpy.promote_types(dtype, arg.dtype)
+
+        outs = kwargs.pop('out', None)
+        axes = kwargs.pop('axes', None)
+        axis = kwargs.pop('axis', None)
+        keepdims = kwargs.pop('keepdims', False)
+        if len(kwargs) > 0:
+            raise RuntimeError(
+                'Unknown kwargs {}'.format(' '.join(kwargs.keys())))
+
+        if not type(self._signature) == str:
+            raise TypeError('`signature` has to be of type string')
+
+        if outs is not None and type(outs) != tuple:
+            if isinstance(outs, cupy.ndarray):
+                outs = (outs,)
+            else:
+                raise TypeError('`outs` must be a tuple or `cupy.ndarray`')
+
+        input_coredimss = self._input_coredimss
+        output_coredimss = self._output_coredimss
+        if outs is not None and type(outs) != tuple:
+            raise TypeError('`outs` must be a tuple')
+        # Axes
+        input_axes, output_axes = _validate_normalize_axes(
+            axes, axis, keepdims, input_coredimss, output_coredimss
+        )
+
+        if len(input_coredimss) != len(args):
+            ValueError(
+                'According to `signature`, `func` requires %d arguments,'
+                ' but %s given' % (len(input_coredimss), len(args)))
+
+        args, dimsizess, loop_output_dims, outs, m_dims = self._get_args_transposed(  # NOQA
+            args, input_axes, input_coredimss,
+            outs, output_axes, output_coredimss)
+
+        # The output shape varies depending on optional dims or not
+        # TODO(ecastill) this only works for one out argument
+        out_shape = tuple([dimsizess[od][0] for od in loop_output_dims]
+                          + [dimsizess[od][0] for od in output_coredimss])
+        if outs is None:
+            outs = (cupy.zeros(out_shape, dtype=ret_dtype),)
```

Comment:
Don't need to initialize with zeros.
```suggestion
            outs = (cupy.empty(out_shape, dtype=ret_dtype),)
```

Reply:
The size of `outs` should be `self._nout`?

Reply:
I think this is for the case where a gufunc returns two arrays, currently we only support returning one array.